ddtrace/internal/llmobs/integrations/openai.py

```diff
@@ -122,94 +123,109 @@ def record_usage(self, span: Span, usage: Dict[str, Any]) -> None:
         if not self.llmobs_enabled:
             return
-        choices = resp.choices
-        now = time.time()
+        if err is not None:
+            attrs_dict = self._llm_error_record(span, kwargs, "chat")
+            self.llm_record(span, attrs_dict)
+            return
         # Note: LLMObs ingest endpoint only accepts a 1:1 prompt-response mapping per record,
         #  so we need to send unique prompt-response records if there are multiple responses (n > 1).
-        for choice in choices:
-            messages = kwargs.get("messages", [])
+        for choice in resp.choices:
             content = getattr(choice.message, "content", None)
             if getattr(choice.message, "function_call", None):
                 content = choice.message.function_call.arguments
             elif getattr(choice.message, "tool_calls", None):
                 content = choice.message.tool_calls.function.arguments
-            # OpenAI only returns the aggregate token count for the entire response if n>1. This means we can only
-            #  provide a rough estimate of the number of tokens used for each completion by taking the average.
-            prompt_tokens = int(resp.usage.prompt_tokens / len(choices))
-            completion_tokens = int(resp.usage.completion_tokens / len(choices))
-            attrs_dict = {
-                "type": "chat",
-                "id": resp.id,
-                "timestamp": resp.created * 1000,
-                "model": resp.model or span.get_tag("openai.request.model"),
-                "model_provider": "openai",
-                "input": {
-                    "messages": [{"content": str(m.get("content", "")), "role": m.get("role", "")} for m in messages],
-                    "temperature": kwargs.get("temperature"),
-                    "max_tokens": kwargs.get("max_tokens"),
-                    "prompt_tokens": [prompt_tokens],
-                },
-                "output": {
-                    "completions": [
-                        {
-                            "content": str(content),
-                            "role": choice.message.role,
-                        }
-                    ],
-                    "durations": [now - span.start],
-                    "completion_tokens": [completion_tokens],
-                    "total_tokens": [prompt_tokens + completion_tokens],
-                    "rate_limit_requests": [span.get_metric("openai.organization.ratelimit.requests.limit")],
-                    "rate_limit_tokens": [span.get_metric("openai.organization.ratelimit.tokens.limit")],
-                },
-            }
+            attrs_dict = self._llmobs_record(span, kwargs, resp, "chat")
+            attrs_dict["output"]["completions"] = [{"content": str(content), "role": choice.message.role}]
             self.llm_record(span, attrs_dict)
+
+    def _llm_error_record(self, span: Span, kwargs: Dict[str, Any], record_type: str) -> Dict[str, Any]:
```

Comment:
Could these two record-generating methods be merged into one, that takes both the response and error arg, and depending on which is set, conditionally add things to `attrs_dict`? I'm thinking it might be better for readability to see at a quick glance what is added, or not added, to the attributes dict in the case of an error or a proper response, and would be DRYed up a bit

On the other hand, having it the way you do now makes the separate uses easier to find in code, and shows that errored records are handled, if someone were to take a quick glance through the code.

Just wanted to see if you had thoughts here, but definitely not blocking, the logic here in general seems right to me!

Reply:
Fair point, I see the pros/cons of both approaches and don't have a strong preference. I'll merge the two helpers into one